testenv/aws/utils.go

```diff
@@ -0,0 +1,291 @@
+	})
+	if err != nil {
+		return fmt.Errorf("failed to delete bucket: %w", err)
+	}
+
+	return nil
+}
+
+// Get stack status by stack name.
+func (e *AWSEnv) getStackStatus(ctx context.Context) (cloudformationtypes.StackStatus, error) {
+	stacks, err := e.client.DescribeStacks(
+		ctx,
+		&cloudformation.DescribeStacksInput{
+			StackName: &e.stackName,
+		},
+	)
+	if err != nil {
+		return "", fmt.Errorf("failed to describe stack: %w", err)
+	}
+
+	if len(stacks.Stacks) != 1 {
+		return "", errors.New("failed to find stack")
+	}
+
+	return stacks.Stacks[0].StackStatus, nil
+}
+
+func (e *AWSEnv) getServer(ctx context.Context) (*Server, error) {
+	// Get list of stack resources
+	resources, err := e.client.ListStackResources(
+		ctx,
+		&cloudformation.ListStackResourcesInput{
+			StackName: &e.stackName,
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list stack resources: %w", err)
+	}
+
+	// Get VMClarity Server EC2 instance ID
+	var serverInstanceID string
+	for _, resource := range resources.StackResourceSummaries {
+		if *resource.ResourceType == "AWS::EC2::Instance" {
+			serverInstanceID = *resource.PhysicalResourceId
+
+			break
+		}
+	}
+
+	publicIP, err := e.getServerPublicIP(ctx, serverInstanceID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get public IP address for instance with %s ID: %w", serverInstanceID, err)
+	}
+
+	return &Server{
+		InstanceID: serverInstanceID,
+		PublicIP:   publicIP,
+	}, nil
+}
+
+// Check if the infrastructure is ready.
+func (e *AWSEnv) infrastructureReady(ctx context.Context) (bool, error) {
+	// Get stack status
+	// If the stack status is not CREATE_COMPLETE, then the infrastructure is not ready
+	stackStatus, err := e.getStackStatus(ctx)
+	if err != nil {
+		return false, fmt.Errorf("failed to get stack status: %w", err)
+	}
+	if stackStatus != cloudformationtypes.StackStatusCreateComplete {
+		return false, nil
+	}
+
+	// Get test asset status
+	// If the test asset status is not running, then the infrastructure are not ready
+	testAssetStatus, err := e.getEC2InstanceStatus(ctx, e.testAsset.InstanceID)
+	if err != nil {
+		return false, fmt.Errorf("failed to get test instance status: %w", err)
+	}
+	if testAssetStatus != ec2types.InstanceStateNameRunning {
+		return false, nil
+	}
+
+	// Get VMClarity Server EC2 instance status
+	// If the server status is not running, then the infrastructure are not ready
+	e.server, err = e.getServer(ctx)
+	if err != nil {
+		return false, fmt.Errorf("failed to get VMClarity server: %w", err)
+	}
+
+	status, err := e.getEC2InstanceStatus(ctx, e.server.InstanceID)
+	if err != nil {
+		return false, fmt.Errorf("failed to get server instance status: %w", err)
+	}
+	if status != ec2types.InstanceStateNameRunning {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// Get EC2 instance status by instance ID.
+func (e *AWSEnv) getEC2InstanceStatus(ctx context.Context, instanceID string) (ec2types.InstanceStateName, error) {
```

Comment:
if this is only used to check if the instance is ready, maybe this could be changed into `isEC2InstanceReady`